### PR TITLE
tools/runqslower.py: Fix raw tracepoint code

### PR DESCRIPTION
**Pull request by bodgergely**
Closes: #2588
It is was observed that the same PID can show up many
times to many unrelated processes. The exact cause of
this is unknown but highly likely (speculation) a race condition.

Removing the default (if supported) raw tracepoint option along
with the raw tracepoint code.

### CODE REVIEW

**Comment by bodgergely** [comment]
I tested and using the kprobe code instead of the raw tracepoint fixes the issue.

**Comment by yonghong-song** [comment]
I would like to know what is the exact reason for raw tracepoint having issues. If you do not have time, I will take a look some time next week.

**Comment by bodgergely** [comment]
@yonghong-song Actually I got quite interested as well, so I will take a look now.

**Comment by bodgergely** [comment]
It is definitely a race condition. I was able to reduce the amount of times the race happens by moving code around etc (the bpf_probe_reads mostly) but I get a sense we have a deeper dependency/design issue here (either the linux kernel raw tracepoint mechanism or the bcc itself). Still needs investigating whether it is possible to guarantee race elimination at all with the raw tp method.

**Comment by bodgergely** [comment]
I think the  `bpf_get_current_comm(&data.task, sizeof(data.task));` is the culprit here. It is possible that is is operating on a more recently scheduled task than the one we captured by `next`. That explains why we see another task's name associated with the pid. 
I used 
```
    bpf_trace_printk("prev_pid: %u,prev_comm: %s\\n", prev->pid, prev->comm);
    bpf_trace_printk("next_pid: %u,next_comm: %s\\n", next->pid, next->comm);

```
where I can simply log out the current task's name without using `bpf_get_current_comm` and the process name-pid pairs are correct when looking into `trace_pipe`
We should somehow extract the name of the task atomically along with the pid.

**Comment by bodgergely** [comment]
Using the
```
bpf_probe_read_str(&data.task, sizeof(data.task), next->comm);
```
instead of 
```
bpf_get_current_comm(&data.task, sizeof(data.task));
```
eliminates most of the erroneous results.
I updated the PR.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
Thanks! The change makes sense. The `pid` we got is from `next` pid. The `comm` should be the one from `next`, not from current.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by bodgergely** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by bodgergely** [comment]
Should I squash the 2 commits into a single commit?

File: tools/runqslower.py [hunk 1]
```diff
@@ -161,25 +161,23 @@
...
     // TP_PROTO(bool preempt, struct task_struct *prev, struct task_struct *next)
-    struct task_struct *prev = (struct task_struct *)ctx->args[1];
-    struct task_struct *next= (struct task_struct *)ctx->args[2];
-    u32 pid, tgid;
+    u32 pid;
     long state;
+    struct task_struct *next = (struct task_struct *)ctx->args[2];
+    struct task_struct *prev = (struct task_struct *)ctx->args[1];
+
+    bpf_probe_read(&pid, sizeof(next->pid), &next->pid);
```
Comment (yonghong-song) [inline tools/runqslower.py]: This does not sound right. The pid could be overwritten by later at line 174. I suggest keeping code as is, just removing the dead code at line 172 to bpf_probe_read value for tgid.

**Comment by yonghong-song** [comment]
You need to fix a problem I just pointed out. Normally I can squash the commits when I apply. But since you need to change codes any way, please go ahead to squash. Thanks!

File: tools/runqslower.py (hunk 1 above)
Comment (bodgergely) [reply tools/runqslower.py]: Understood, you are correct about line 174 overwriting the pid. I will fix that. Might as well move the line 169 after the block:
```
    if (state == TASK_RUNNING) {
        bpf_probe_read(&pid, sizeof(prev->pid), &prev->pid);
        if (!(FILTER_PID || pid == 0)) {
            u64 ts = bpf_ktime_get_ns();
            start.update(&pid, &ts);
        }
    }

```
> just removing the dead code at line 172 to bpf_probe_read value for tgid.
Line 172:
```
bpf_probe_read(&state, sizeof(long), (const void *)&prev->state);
```
The above does not seem dead code to me. If I remove line 172, the `state` variable will stay undefined. I might missing something here.

Also the `tgid` is not used at all. So why introduce is back?

Comment (yonghong-song) [reply tools/runqslower.py]: I mean tgid code. The 'state' bpf_probe_read certainly useful. The line number sometimes refers to old, sometimes refer to new code, so maybe confusing.

Comment (yonghong-song) [reply tools/runqslower.py]: I think just remove dead code and use bpf_probe_read_str() change should be good enough.

**Comment by bodgergely** [comment]
Did the fix as we agreed.

**Comment by bodgergely** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

**Comment by yonghong-song** [comment]
Merged this patch. The change looks good. But current runqslower.py seems filtering in bpf program based on task id instead of user visible pid. 

Suppose pid is the task id and tgid is the actual process id,
the following code
```
bpf_text = bpf_text.replace('FILTER_PID', 'pid != %s' % args.pid)
```
probably should be
```
bpf_text = bpf_text.replace('FILTER_PID', 'tgid != %s' % args.pid)
```

Did you run into any issues with this? We probably should fix this issue as well.
The whole bpf program 'pid' thing can be then removed, `tgid` will be used then.

**Comment by bodgergely** [comment]
Thanks, I created an Issue https://github.com/iovisor/bcc/issues/2607
regarding your comment to look into it later.

**Pull request merged**